### DISCUSSION TITLE
Add an ability to have multiple root elements per page

### DISCUSSION THREAD

**Discussion by drupsys** [vercel/next.js#38759] [Ideas]
### Describe the feature you'd like to request

I want to be able to create a page that has many root elements, where each root attaches to the specified div so that I could load nextjs into a legacy application where only parts of the (legacy) page would be replaced with react components.

Microfrontend functionality that nextjs currently provides is appreciated but not very useful for those of us that are migrating real world legacy applications to react. The main issue is that in order to migrate something to react (in nextjs) you have to rewrite the entire page, this is not realistic, each page in the application that I work on is shockingly complex. We want to use react whenever we write any new functionality but realisticly this new functionality will only ever touch 5% of the page (at most) so noone will ever give us time to rewrite 95% of the page in react just to get the 5% done.

### Describe the solution you'd like

This is by no means the only way nor neceseraly correct way, but one thing you could do is provide a new component, lets say `Snippet`, that we could use in our pages, e.g.

in `./pages/page.tsx`
```tsx
import type { NextPage, Snippet } from 'next';
import { Fragment } from 'react';

const Home: NextPage = () => {
  return <Fragment>
    <Snippet target='div-id-1'>
      content of snippet 1
    </Snippet>
    <Snippet target='div-id-2'>
      content of snippet 2
    </Snippet>
    <Snippet target='div-id-...'>
      content of snippet ...
    </Snippet>
  </Fragment>;
};
```

P.S. maybe also add an option to attach snippet as a shadow dom, in most cases that would resolve a number of issues with styling that is common when doing this sort of stuff, but that is a nice to have, there are other ways to deal with styling issues.

Then in your legacy application you would need to load the assets associated with the above page, I don't think there is currently a way to easily just load the assets without the initial html template so there would need to be something for that

```html
<div id="div-id-1"></div>
<div id="div-id-2"></div>
<div id="div-id-..."></div>

// assuming nextjs is hosted on `/micro/temp-frontend` path
// this should load the assets of `./pages/page.tsx`
<script src="/micro/temp-frontend/page.js"></script>
```

### Describe alternatives you've considered

I currently maintain my own webpack setup to allow the above.

(to be clear this means I can't use nextjs and I would really like to it would make my life way easier if I could)

**Comment by drupsys** [comment]
Any opinions about this from the maintainers? Any information would be appreciated event if it's just "we are not looking to add this functionality" because then I can at least make my own plans.

**Comment by Soundvessel** [comment]
I am in a similar situation of working on a new react front end for a ColdFusion based site. Would love to use the nextjs ecosystem because eventually all the updated components would begin to merge into what NextJS does best on the entire page.

**Comment by lucksp** [comment]
following.  what solutions have been found so far?
